### PR TITLE
Add space POST endpoint

### PR DESCRIPTION
**Pull request by mattnibs**
Add the ability to create a new empty space via the space post
endpoint. The space post endpoint requires either a space name or
a space data dir (additionally both can be provided).

If a space data dir is specified, with no space name, a new space
is created given the name of data dir's base directory. If the name
collides with an exisiting space, a unique name is generated and
returned in the request response.

### CODE REVIEW

File: pkg/fs/unique.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// UniquePath determines a unique path given the desired name and parent
+// directory. For instanace of the desired path in /usr/mypath but such a path
```
Comment (nwt) [inline pkg/fs/unique.go]:
```suggestion
// directory. For instance if the desired path is /usr/mypath but such a path
```

File: pkg/fs/unique.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	matches, err := filepath.Glob(filepath.Join(parent, base+"*"+ext))
+	if err != nil {
+		return "", err
+	}
+	if len(matches) > 0 {
+		slice := sort.StringSlice(matches)
+		slice.Sort()
+		for i := 1; true; i++ {
+			name = fmt.Sprintf("%s_%02d%s", base, i, ext)
+			if n := slice.Search(base); n == slice.Len() {
```
Comment (nwt) [inline pkg/fs/unique.go]:
```suggestion
			if slice.Search(base) == slice.Len() {
```

File: pkg/fs/unique.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+func UniquePath(parent, name string) (string, error) {
+	ext := filepath.Ext(name)
+	base := strings.TrimSuffix(name, ext)
+	matches, err := filepath.Glob(filepath.Join(parent, base+"*"+ext))
+	if err != nil {
+		return "", err
+	}
+	if len(matches) > 0 {
+		slice := sort.StringSlice(matches)
+		slice.Sort()
```
Comment (nwt) [inline pkg/fs/unique.go]:
```suggestion
		sort.Strings(matches)
```

File: zqd/api/api.go [hunk 4]
```diff
@@ -80,6 +80,16 @@ type StatusResponse struct {
...
+type SpacePostRequest struct {
+	Name    string `json:"name"`
+	DataDir string `json:"data_dir"`
+}
+
+type SpacePostResponse struct {
+	Name    string `json:"name"`
+	DataDir string `json:"data_dir"`
+}
+
```
Comment (nwt) [inline zqd/api/api.go]:
```suggestion
type SpacePostResponse SpacePostRequest
```

File: pkg/fs/unique.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+// UniquePath determines a unique path given the desired name and parent
+// directory. For instanace of the desired path in /usr/mypath but such a path
+// already exists, the path /usr/mypath_01 is returned (and so on).
+// File extensions are respected and the unique number is appended before any
+// extensions (e.g. mypath.txt -> mypath_01.txt).
+func UniquePath(parent, name string) (string, error) {
+	ext := filepath.Ext(name)
+	base := strings.TrimSuffix(name, ext)
+	matches, err := filepath.Glob(filepath.Join(parent, base+"*"+ext))
```
Comment (nwt) [inline pkg/fs/unique.go]: I don't think this does the right thing if `name` doesn't exist but `base+"x"+ext` does. I suggest doing something similar to what ioutil.TempDir does.

Comment (mattnibs) [reply pkg/fs/unique.go]: ok

**Comment by mattnibs** [comment]
> It would be nice if the file path were optional and set to the path of $root/name if absent.

@mccanne this is what it does if `data_dir` is not specified in the post request.

**Pull request merged**